R/PackageDependencyReporter.R

```diff
@@ -114,11 +123,87 @@ PackageDependencyReporter <- R6::R6Class(
             
             dependencyList <- Filter(function(x){!is.null(x)}, dependencyList)
             
-            edges <- data.table::rbindlist(lapply(names(dependencyList), function(pkgN){data.table::data.table(SOURCE = rep(pkgN,length(dependencyList[[pkgN]]))
-                                                                                                               ,TARGET = dependencyList[[pkgN]])
-            }))
+            edges <- data.table::rbindlist(lapply(
+                names(dependencyList), 
+                function(pkgN){
+                    data.table::data.table(SOURCE = rep(pkgN,length(dependencyList[[pkgN]])), 
+                                           TARGET = dependencyList[[pkgN]])
+                    }
+                ))
+            
+            private$cache$edges <- edges
+            
+            # Get and save nodes
+            nodes = data.table::data.table(node = unique(c(self$edges[, SOURCE], self$edges[, TARGET])))
+            private$cache$nodes <- nodes
+            
+            return(list(edges = edges, nodes = nodes))
+        },
+        
+        calculate_all_metrics = function(...) {
+            # Check if we need to re-extract network
+            private$parse_extract_args(list(...))
+            
+            metricsList <- list()
+            
+            # Calculate network measures
+            metricsList <- c(metricsList, self$calculate_network_measures())
             
-            return(edges)
+            return(metricsList)
+        }
+        
+        
+    ),
+    
+    active = list(
+        nodes = function(){
+            if (is.null(private$cache$nodes)){
+                log_info("Calling extract_network() with default arguments...")
+                invisible(self$extract_network())
+            }
+            return(private$cache$nodes)
+        },
+        edges = function(){
+            if (is.null(private$cache$edges)){
+                log_info("Calling extract_network() with default arguments...")
+                invisible(self$extract_network())
+            }
+            return(private$cache$edges)
+        }
+    ),
+    
+    private = list(
+        recursive_dependencies = function(package, db, depTypes, ignorePackages, seenPackages = NULL) {
+            
+            # Package is blacklisted by ignorePackages, stop searching
+            if (package %in% ignorePackages){
+                return(c(seenPackages, package))
```

Comment:
I don't like this `if - else if else` pattern for this case. Since you're returning early, the `else if` is unnecessary complexity. Consider this rewrite:

```
# Case 1 - package is blacklisted by ignorePackages
if (package %in% ignorePackages){
    return(c(seenPackages, package))
}

# Case 2 - we've already seen this package
if (package %in% seenPackages){
    return(seenPackages)
}

# Case 3 - unseen package on the search frontier
[the_rest_of_the_code]
```

I think that this case structure is a cleaner way to write recursive code that also makes the search logic a bit easier to figure out